app/internal/statistics.py

```diff
@@ -0,0 +1,393 @@
+        get_date_filter_between_dates(start, end)).all()
+    if events_duration_statistics[0][0]:
+        return EventsDurationStatistics(
+            shortest_event=round(events_duration_statistics[0][0]),
+            longest_event=round(events_duration_statistics[0][1]),
+            average_event=round(events_duration_statistics[0][2]))
+    return EventsDurationStatistics(
+        shortest_event=0, longest_event=0, average_event=0)
+
+
+def get_events_duration_statistics(
+        db: Session, userid: int, start: datetime.datetime,
+        end: datetime.datetime) -> Dict[str, Dict[str, float]]:
+    """ calculate statistics for events durations relevant for
+        the requested user and the requested date range.
+        all logic is performed in the db, while the rest is in the code.
+
+    Args:
+        db: db session.
+        userid: requested user id number for statistics.
+        start: start of date range.
+        end: end of date range.
+
+    Returns: json with
+        events_duration_statistics: (for requested date range, in minutes).
+            shortest_event: shortest event the user has.
+            longest_event: longest event the user has.
+            average_event: average event the user has.
+    """
+    events_duration_statistics = get_events_duration_statistics_from_db(
+        db, userid, start, end)
+    return {
+        "events_duration_statistics": {
+            "shortest_event": events_duration_statistics.shortest_event,
+            "longest_event": events_duration_statistics.longest_event,
+            "average_event": events_duration_statistics.average_event}}
+
+
+def get_participants_statistics(
+        db: Session, userid: int, start: datetime.datetime,
+        end: datetime.datetime) -> Dict[str, Dict[str, Union[str, int]]]:
+    """ calculate statistics for events participants relevant for
+        the requested user and the requested date range.
+        part of the logic is performed in the db,
+        while the rest is in the code.
+
+    Args:
+        db: db session.
+        userid: requested user id number for statistics.
+        start: start of date range.
+        end: end of date range.
+
+    Returns: json with
+        max_events: maximum number of events the user has
+            with same participant.
+        participant_name: relevant participant name.
+    """
+    by_user_id = UserEvent.user_id == userid
+    by_not_user_id = UserEvent.user_id != userid
+    user_to_event = (UserEvent, UserEvent.event_id == Event.id)
+    participant_count = func.count(UserEvent.user_id)
+    subquery = db.query(
+        Event.id).join(user_to_event).filter(by_user_id).filter(
+        get_date_filter_between_dates(start, end)).subquery()
+    event_participants = db.query(
+        UserEvent.user_id, participant_count).filter(
+        by_not_user_id).filter(
+        UserEvent.event_id.in_(subquery)).group_by(
+        UserEvent.user_id).order_by(participant_count.desc()).first()
```

Comment:
Use pre-commit hooks to automatically fix the formatting here (there is an explanation in the README)